### PR TITLE
Extend tensor API

### PR DESCRIPTION
**Pull request by ilyachur**
### Details:
 - Should be completed after https://github.com/openvinotoolkit/openvino/pull/15696

### Tickets:
 - *ticket-id*

### CODE REVIEW

File: src/core/src/runtime/ov_tensor.cpp [hunk 1]
```diff
@@ -105,6 +118,119 @@ Shape Tensor::get_shape() const {
...
 
+void Tensor::copy_to(const ov::Tensor& dst) const {
+    OPENVINO_ASSERT(dst.get_element_type() == get_element_type(),
+                    "Tensor element types are not equal. (src: ",
+                    get_element_type(),
+                    " != dst: ",
+                    dst.get_element_type(),
+                    ")");
+    OPENVINO_ASSERT(dst.get_shape() == get_shape(),
+                    "Tensor shapes are not equal. (src: ",
```
Comment (ilya-lavrenov) [inline src/core/src/runtime/ov_tensor.cpp:129]: I suppose `dst` tensor can be reallocated in shapes don't match. 
E.g. I can create empty tensor and copy to it:
```c++
ov::Tensor dst, src = get_src_tensor();
src.copy_to(dst);
```

File: src/core/include/openvino/runtime/tensor.hpp [hunk 2]
```diff
@@ -137,6 +154,13 @@ class OPENVINO_API Tensor {
      */
     Shape get_shape() const;
 
+    /**
+     * @brief Copy tensor
+     *
+     * @param dst destination tensor, should have the same element type and shape
+     */
```
Comment (jiwaszki) [inline src/core/include/openvino/runtime/tensor.hpp:161]:
```suggestion
    /**
     * @brief Copy tensor's data into another tensor. Destination tensor should have the same element type and shape.
     *
     * @param dst destination tensor.
     */
```
I think comment should be more descriptive about action.

What about strides if `copy_to` is used?

Comment (ilyachur) [reply src/core/include/openvino/runtime/tensor.hpp:161]: Strides can be different. Also I will create a ticket to extend tensor API in python

File: src/core/tests/ov_tensor_test.cpp [hunk 3]
```diff
@@ -303,3 +303,128 @@ TEST_F(OVTensorTest, readRangeRoiBlob) {
...
+    ov::Tensor src_tensor{full_tensor, {0, 2, 0}, {3, 4, 2}};
+    ov::Tensor dst_tensor{ov::element::i32, {3, 4, 8}};
+    {
+        const auto origPtr = full_tensor.data<int32_t>();
+        ASSERT_NE(nullptr, origPtr);
+        for (size_t i = 0; i < full_tensor.get_size(); ++i) {
+            origPtr[i] = static_cast<int32_t>(i);
+        }
+    }
+    ov::Tensor with_strides{ov::element::i32, ov::Shape{3, 2, 2}, dst_tensor.data(), ov::Strides{32 * 4, 6 * 4, 2 * 4}};
```
Comment (ilya-lavrenov) [inline src/core/tests/ov_tensor_test.cpp:426]: should we generalize this test to be a parametrized tests where parameters are strides on different dimensions?

Comment (ilyachur) [reply src/core/tests/ov_tensor_test.cpp]: done

File: src/core/tests/ov_tensor_test.cpp [hunk 4]
```diff
@@ -281,3 +303,94 @@ TEST_F(OVTensorTest, readRangeRoiBlob) {
...
+        for (size_t i = 0; i < source_vec.size(); i++) {
+            EXPECT_EQ(source_vec[i], dest_vec[i]);
+        }
+    }
+}
+
+// clang-format off
+INSTANTIATE_TEST_SUITE_P(copy_tests,
+                         OVTensorTestCopy,
+                         ::testing::Values(
```
Comment (ilya-lavrenov) [inline src/core/tests/ov_tensor_test.cpp:378]: should we also add an element_type as a parameter as well? E.g. to check different sizes of elements

File: src/core/src/runtime/ov_tensor.cpp [hunk 5]
```diff
@@ -113,6 +126,123 @@ Shape Tensor::get_shape() const {
     OV_TENSOR_STATEMENT({ return _impl->getTensorDesc().getBlockingDesc().getBlockDims(); });
 }
 
+void Tensor::copy_to(ov::Tensor& dst) const {
+    OPENVINO_ASSERT(dst.get_element_type() == get_element_type(),
```
Comment (ilya-lavrenov) [inline src/core/src/runtime/ov_tensor.cpp:130]: how does it work for:
- RemoteTensor::copy_to(Tensor)
- RemoteTensor::copy_to(RemoteTensor)
- Tensor::copy_to(RemoteTensor)

Comment (ilyachur) [reply src/core/src/runtime/ov_tensor.cpp:130]: I will add a check that tensors shouldn't be remote

**Review by ilya-lavrenov — APPROVED** [review]

**Pull request merged**